compiler/hash-typecheck/src/exhaustiveness/construct.rs

```diff
@@ -0,0 +1,324 @@
+    List(List),
+    /// Wildcard pattern.
+    Wildcard,
+    /// Or-pattern.
+    Or,
+    /// Stands for constructors that are not seen in the matrix, as explained in
+    /// the documentation for [super::wildcard::SplitWildcard].
+    Missing,
+    /// Declared as non-exhaustive
+    NonExhaustive,
+}
+
+impl Constructor {
+    /// Check if the [Constructor] is a wildcard.
+    pub(super) fn is_wildcard(&self) -> bool {
+        matches!(self, Constructor::Wildcard)
+    }
+
+    /// Try and convert the [Constructor] into a [IntRange].
+    pub fn as_int_range(&self) -> Option<&IntRange> {
+        match self {
+            Constructor::IntRange(range) => Some(range),
+            _ => None,
+        }
+    }
+
+    /// Try and convert the [Constructor] into a [List].
+    pub fn as_list(&self) -> Option<&List> {
+        match self {
+            Constructor::List(list) => Some(list),
+            _ => None,
+        }
+    }
+}
+
+pub struct ConstructorOps<'tc> {
+    storage: StorageRef<'tc>,
+}
+
+impl<'tc> AccessToStorage for ConstructorOps<'tc> {
+    fn storages(&self) -> StorageRef {
+        self.storage.storages()
+    }
+}
+
+impl<'tc> ConstructorOps<'tc> {
+    /// Create a new [ConstructorOps].
+    pub fn new(storage: StorageRef<'tc>) -> Self {
+        Self { storage }
+    }
+
+    /// Create a [SourceLocation] from a provided [Span].
+    pub fn location(&self, span: Span) -> SourceLocation {
+        SourceLocation { span, source_id: self.local_storage().current_source() }
+    }
+
+    /// Compute the `arity` of this [Constructor].
+    pub(crate) fn arity(&self, ctx: PatCtx, ctor: ConstructorId) -> usize {
+        let reader = self.reader();
+
+        match reader.get_ctor(ctor) {
+            Constructor::Single | Constructor::Variant(_) => {
+                // we need to get term from the context here...
+                //
+                // if it a tuple, get the length and that is the arity
+                // if it is a struct or enum, then we get that variant and
+                // we can count the fields from that variant or struct.
+                let reader = self.reader();
+
+                match reader.get_term(ctx.ty) {
+                    Term::Level1(Level1Term::Tuple(TupleTy { members })) => {
+                        reader.get_params(*members).len()
+                    }
+                    Term::Level1(Level1Term::NominalDef(def)) => {
+                        match reader.get_nominal_def(*def) {
+                            NominalDef::Struct(struct_def) => match struct_def.fields {
+                                StructFields::Explicit(params) => reader.get_params(params).len(),
+                                StructFields::Opaque => 0,
+                            },
+                            // @@Remove: when enums are no longer represented as thi
+                            NominalDef::Enum(_) => unreachable!(),
+                        }
+                    }
```

Comment:
This could possibly be another thing that can benefit from the oracle, maybe `count_params_of_ty` or something.

Reply:
This would be a very beneficial utility, and possibly things like `as_struct`, `as_enum_variant`...